File: src/jeeves/core/paths.py
"""XDG-compliant path resolution for Jeeves.

This module provides standardized paths for Jeeves data following XDG Base Directory
Specification via platformdirs.

Directory structure:
    ~/.local/share/jeeves/           # JEEVES_DATA_DIR
    ├── repos/                       # Bare or full clones (shared)
    │   └── {owner}/{repo}/
    ├── worktrees/                   # Git worktrees (one per issue)
    │   └── {owner}/{repo}/issue-{N}/
    ├── issues/                      # Issue state (decoupled from worktree)
    │   └── {owner}/{repo}/{N}/
    │       ├── issue.json
    │       ├── progress.txt
    │       └── .runs/
    └── config.json                  # Global settings
"""

import os
from pathlib import Path
from typing import Optional

try:
    import platformdirs
except ImportError:
    platformdirs = None  # type: ignore


def get_data_dir() -> Path:
    """Get the Jeeves data directory.

    Uses XDG standard paths via platformdirs:
    - Linux: ~/.local/share/jeeves
    - macOS: ~/Library/Application Support/jeeves
    - Windows: ~/AppData/Local/jeeves

    Can be overridden with JEEVES_DATA_DIR environment variable.

    Returns:
        Path to the data directory.
    """
    env_dir = os.environ.get("JEEVES_DATA_DIR")
    if env_dir:
        return Path(env_dir).expanduser().resolve()

    if platformdirs is not None:
        return Path(platformdirs.user_data_dir("jeeves", appauthor=False))

    # Fallback if platformdirs not installed
    if os.name == "nt":
        # Windows
        base = os.environ.get("LOCALAPPDATA", os.path.expanduser("~"))
        return Path(base) / "jeeves"
    elif os.uname().sysname == "Darwin":
        # macOS
        return Path.home() / "Library" / "Application Support" / "jeeves"
    else:
        # Linux/Unix - XDG_DATA_HOME
        xdg_data = os.environ.get("XDG_DATA_HOME")
        if xdg_data:
            return Path(xdg_data) / "jeeves"
        return Path.home() / ".local" / "share" / "jeeves"


def get_repos_dir() -> Path:
    """Get the directory for repository clones.

    Returns:
        Path to repos directory (~/.local/share/jeeves/repos).
    """
    return get_data_dir() / "repos"


def get_worktrees_dir() -> Path:
    """Get the directory for git worktrees.

    Returns:
        Path to worktrees directory (~/.local/share/jeeves/worktrees).
    """
    return get_data_dir() / "worktrees"


def get_issues_dir() -> Path:
    """Get the directory for issue state files.

    Returns:
        Path to issues directory (~/.local/share/jeeves/issues).
    """
    return get_data_dir() / "issues"


def get_config_file() -> Path:
    """Get the path to the global config file.

    Returns:
        Path to config.json.
    """
    return get_data_dir() / "config.json"


def get_repo_path(owner: str, repo: str) -> Path:
    """Get the path to a specific repository clone.

    Args:
        owner: Repository owner (e.g., 'anthropics').
        repo: Repository name (e.g., 'claude-code').

    Returns:
        Path to the repository directory.
    """
    return get_repos_dir() / owner / repo


def get_worktree_path(owner: str, repo: str, issue_number: int) -> Path:
    """Get the path to a worktree for a specific issue.

    Args:
        owner: Repository owner.
        repo: Repository name.
        issue_number: Issue number.

    Returns:
        Path to the worktree directory.
    """
    return get_worktrees_dir() / owner / repo / f"issue-{issue_number}"


def get_issue_state_dir(owner: str, repo: str, issue_number: int) -> Path:
    """Get the path to the state directory for a specific issue.

    Args:
        owner: Repository owner.
        repo: Repository name.
        issue_number: Issue number.

    Returns:
        Path to the issue state directory.
    """
    return get_issues_dir() / owner / repo / str(issue_number)


def parse_repo_spec(spec: str) -> tuple[str, str]:
    """Parse a repository specification into owner and repo.

    Accepts formats:
    - owner/repo
    - https://github.com/owner/repo
    - https://github.com/owner/repo.git
    - git@github.com:owner/repo.git

    Args:
        spec: Repository specification string.

    Returns:
        Tuple of (owner, repo).

    Raises:
        ValueError: If the spec cannot be parsed.
    """
    spec = spec.strip()

    # Simple owner/repo format
    if "/" in spec and not spec.startswith(("http://", "https://", "git@")):
        parts = spec.split("/")
        if len(parts) == 2:
            owner, repo = parts
            repo = repo.removesuffix(".git")
            if owner and repo:
                return owner, repo

    # HTTPS URL format
    if spec.startswith(("http://", "https://")):
        # Remove protocol and domain
        path = spec.split("github.com/")[-1] if "github.com/" in spec else ""
        path = path.removesuffix(".git").rstrip("/")
        parts = path.split("/")
        if len(parts) >= 2:
            return parts[0], parts[1]

    # SSH format: git@github.com:owner/repo.git
    if spec.startswith("git@"):
        path = spec.split(":")[-1] if ":" in spec else ""
        path = path.removesuffix(".git")
        parts = path.split("/")
        if len(parts) == 2:
            return parts[0], parts[1]

    raise ValueError(
        f"Invalid repository specification: {spec!r}. "
        "Expected format: owner/repo, https://github.com/owner/repo, "
        "or git@github.com:owner/repo.git"
    )


def parse_issue_ref(ref: str) -> tuple[str, str, int]:
    """Parse an issue reference into owner, repo, and issue number.

    Accepts formats:
    - owner/repo#123
    - #123 (requires context repo)
    - https://github.com/owner/repo/issues/123

    Args:
        ref: Issue reference string.

    Returns:
        Tuple of (owner, repo, issue_number).

    Raises:
        ValueError: If the ref cannot be parsed.
    """
    ref = ref.strip()

    # owner/repo#123 format
    if "#" in ref:
        repo_part, issue_part = ref.rsplit("#", 1)
        try:
            issue_number = int(issue_part)
        except ValueError:
            raise ValueError(f"Invalid issue number: {issue_part!r}")

        if "/" in repo_part:
            owner, repo = parse_repo_spec(repo_part)
            return owner, repo, issue_number
        else:
            raise ValueError(
                f"Cannot determine repository from {ref!r}. "
                "Use format: owner/repo#123"
            )

    # GitHub issue URL format
    if "github.com" in ref and "/issues/" in ref:
        # https://github.com/owner/repo/issues/123
        parts = ref.split("/issues/")
        if len(parts) == 2:
            repo_url, issue_str = parts
            try:
                issue_number = int(issue_str.split("/")[0].split("?")[0])
            except ValueError:
                raise ValueError(f"Invalid issue URL: {ref!r}")
            owner, repo = parse_repo_spec(repo_url)
            return owner, repo, issue_number

    # Just a number - try to parse as issue number
    try:
        issue_number = int(ref)
        raise ValueError(
            f"Issue number {issue_number} requires a repository. "
            "Use format: owner/repo#123"
        )
    except ValueError:
        if ref.isdigit():
            raise
        raise ValueError(
            f"Invalid issue reference: {ref!r}. "
            "Expected format: owner/repo#123 or GitHub issue URL"
        )


def ensure_directory(path: Path) -> Path:
    """Ensure a directory exists, creating it if necessary.

    Args:
        path: Path to the directory.

    Returns:
        The path (for chaining).
    """
    path.mkdir(parents=True, exist_ok=True)
    return path


def is_legacy_mode() -> bool:
    """Legacy mode removed; keep for backward compatibility."""
    return False

File: packages/core/src/paths.ts
import os from 'node:os';
import path from 'node:path';

export type RepoSpec = Readonly<{ owner: string; repo: string }>;
export type IssueRef = Readonly<{ owner: string; repo: string; issueNumber: number }>;

function expandHome(inputPath: string, homeDir: string): string {
  if (inputPath === '~') return homeDir;
  if (inputPath.startsWith('~/')) return path.join(homeDir, inputPath.slice(2));
  if (inputPath.startsWith('~\\')) return path.join(homeDir, inputPath.slice(2));
  return inputPath;
}

export function resolveDataDir(options?: {
  env?: NodeJS.ProcessEnv;
  platform?: NodeJS.Platform;
  homeDir?: string;
}): string {
  const env = options?.env ?? process.env;
  const platform = options?.platform ?? process.platform;
  const homeDir = options?.homeDir ?? os.homedir();

  const override = env.JEEVES_DATA_DIR;
  if (override && override.trim()) {
    return path.resolve(expandHome(override.trim(), homeDir));
  }

  if (platform === 'win32') {
    const base = env.LOCALAPPDATA ?? path.join(homeDir, 'AppData', 'Local');
    return path.resolve(base, 'jeeves');
  }

  if (platform === 'darwin') {
    return path.resolve(homeDir, 'Library', 'Application Support', 'jeeves');
  }

  const xdgDataHome = env.XDG_DATA_HOME;
  if (xdgDataHome && xdgDataHome.trim()) {
    return path.resolve(xdgDataHome.trim(), 'jeeves');
  }
  return path.resolve(homeDir, '.local', 'share', 'jeeves');
}

export function getDataDir(): string {
  return resolveDataDir();
}

export function getIssuesDir(dataDir: string = getDataDir()): string {
  return path.join(dataDir, 'issues');
}

export function getWorktreesDir(dataDir: string = getDataDir()): string {
  return path.join(dataDir, 'worktrees');
}

/**
 * Legacy issue state directory (pre-2026-02): `${dataDir}/issues/<owner>/<repo>/<issueNumber>`.
 *
 * Jeeves now stores per-issue state inside the issue worktree at `${worktreeDir}/.jeeves`
 * so agent sandboxes and MCP tools can access it without traversing symlinks outside cwd.
 */
export function getLegacyIssueStateDir(
  owner: string,
  repo: string,
  issueNumber: number,
  dataDir: string = getDataDir(),
): string {
  return path.join(getIssuesDir(dataDir), owner, repo, String(issueNumber));
}

export function getIssueStateDir(
  owner: string,
  repo: string,
  issueNumber: number,
  dataDir: string = getDataDir(),
): string {
  return path.join(getWorktreePath(owner, repo, issueNumber, dataDir), '.jeeves');
}

export function getWorktreePath(
  owner: string,
  repo: string,
  issueNumber: number,
  dataDir: string = getDataDir(),
): string {
  return path.join(getWorktreesDir(dataDir), owner, repo, `issue-${issueNumber}`);
}

export function parseRepoSpec(spec: string): RepoSpec {
  const cleaned = spec.trim();
  if (!cleaned) throw new Error('repo spec is required');

  if (cleaned.includes('/') && !cleaned.startsWith('http://') && !cleaned.startsWith('https://') && !cleaned.startsWith('git@')) {
    const parts = cleaned.split('/');
    if (parts.length === 2 && parts[0] && parts[1]) {
      return { owner: parts[0], repo: parts[1].replace(/\.git$/, '') };
    }
  }

  if (cleaned.startsWith('http://') || cleaned.startsWith('https://')) {
    const idx = cleaned.indexOf('github.com/');
    if (idx === -1) throw new Error(`invalid repo spec: ${spec}`);
    const after = cleaned.slice(idx + 'github.com/'.length).replace(/\.git$/, '').replace(/\/$/, '');
    const parts = after.split('/');
    if (parts.length >= 2 && parts[0] && parts[1]) {
      return { owner: parts[0], repo: parts[1] };
    }
  }

  if (cleaned.startsWith('git@')) {
    const segments = cleaned.split(':');
    const after = segments.length > 0 ? segments[segments.length - 1] : '';
    const withoutGit = after.replace(/\.git$/, '');
    const parts = withoutGit.split('/');
    if (parts.length === 2 && parts[0] && parts[1]) {
      return { owner: parts[0], repo: parts[1] };
    }
  }

  throw new Error(
    `invalid repo spec: ${spec}. expected owner/repo, https://github.com/owner/repo, or git@github.com:owner/repo.git`,
  );
}

export function parseIssueRef(ref: string, defaultRepo?: RepoSpec): IssueRef {
  const cleaned = ref.trim();
  if (!cleaned) throw new Error('issue ref is required');

  if (cleaned.startsWith('#')) {
    if (!defaultRepo) {
      throw new Error(`cannot resolve ${ref} without a default repo`);
    }
    const num = Number(cleaned.slice(1));
    if (!Number.isInteger(num) || num <= 0) throw new Error(`invalid issue number: ${cleaned.slice(1)}`);
    return { ...defaultRepo, issueNumber: num };
  }

  if (cleaned.includes('#')) {
    const [repoPart, issuePart] = cleaned.split('#');
    const num = Number(issuePart);
    if (!Number.isInteger(num) || num <= 0) throw new Error(`invalid issue number: ${issuePart}`);
    const parsed = parseRepoSpec(repoPart);
    return { ...parsed, issueNumber: num };
  }

  const match = cleaned.match(/github\.com\/([^/]+)\/([^/]+)\/issues\/(\d+)/);
  if (match) {
    const [, owner, repo, issue] = match;
    return { owner, repo: repo.replace(/\.git$/, ''), issueNumber: Number(issue) };
  }

  if (/^\d+$/.test(cleaned)) {
    if (!defaultRepo) throw new Error(`issue number ${cleaned} requires a repo`);
    return { ...defaultRepo, issueNumber: Number(cleaned) };
  }

  throw new Error(`invalid issue ref: ${ref}`);
}
